mobile_verifier/src/reward_share.rs

```diff
@@ -0,0 +1,297 @@
+
+    async fn owner_shares(
+        &mut self,
+        hotspot_shares: HotspotShares,
+    ) -> Result<(OwnerShares, MissingOwnerShares)> {
+        let mut owner_shares = OwnerShares::new();
+        let mut missing_owner_shares = MissingOwnerShares::new();
+        for (hotspot, share) in hotspot_shares {
+            if let Some(owner) = self.resolve_owner(&hotspot).await? {
+                *owner_shares.entry(owner).or_default() += share;
+            } else {
+                *missing_owner_shares.entry(hotspot).or_default() += share;
+            }
+        }
+        Ok((owner_shares, missing_owner_shares))
+    }
+}
+
+#[async_trait::async_trait]
+impl OwnerResolver for follower::Client<Channel> {
+    async fn resolve_owner(&mut self, address: &PublicKey) -> Result<Option<PublicKey>> {
+        let req = FollowerGatewayReqV1 {
+            address: address.to_vec(),
+        };
+        let res = self.find_gateway(req).await?.into_inner();
+
+        if let Ok(pub_key) = PublicKey::try_from(res.owner) {
+            return Ok(Some(pub_key));
+        }
+
+        Ok(None)
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct GatheredShares {
+    pub shares: Shares,
+    pub invalid_shares: InvalidShares,
+    pub speed_shares: SpeedShares,
+    pub speed_shares_moving_avg: SpeedShareMovingAvgs,
+    pub invalid_speed_shares: InvalidSpeedShares,
+}
+
+impl GatheredShares {
+    pub async fn from_stream(
+        stream: &mut BytesMutStream,
+        after_utc: u64,
+        before_utc: u64,
+    ) -> Result<Self> {
+        let mut shares = Self {
+            shares: Shares::new(),
+            invalid_shares: InvalidShares::new(),
+            speed_shares: SpeedShares::new(),
+            speed_shares_moving_avg: SpeedShareMovingAvgs::default(),
+            invalid_speed_shares: InvalidSpeedShares::new(),
+        };
+
+        while let Some(Ok(msg)) = stream.next().await {
+            // NOTE: This will early exit with an error if we fail to decode
+
+            if let Ok(cell_heartbeat_req_v1) = CellHeartbeatReqV1::decode(msg.clone()) {
+                shares.gather_heartbeat(cell_heartbeat_req_v1, after_utc, before_utc)
+            } else if let Ok(speedtest_req_v1) = SpeedtestReqV1::decode(msg) {
+                shares.gather_speedtest(speedtest_req_v1, after_utc, before_utc)
+            } else {
+                continue;
+            }
+        }
+        Ok(shares)
+    }
+
+    fn gather_speedtest(&mut self, speedtest_req: SpeedtestReqV1, after_utc: u64, before_utc: u64) {
+        // TODO: Better error handling for bad public keys
+        let mut speed_share = match SpeedShare::try_from(speedtest_req) {
+            Ok(share) => share,
+            Err(_) => return,
+        };
+
+        if speed_share.timestamp < after_utc || speed_share.timestamp >= before_utc {
+            speed_share.validity = ShareValidity::HeartbeatOutsideRange;
+            self.invalid_speed_shares.push(speed_share);
```

Comment:
They're written as internal structs if further computation based on them is needed. Anything that's immediately written to a proto struct is not needed any more. 

I don't think this is ideal - for hygiene I think it would be better to write them both to internal structs and then convert them to the serializable formats in the write method. I did this because it was faster to write, but I'd be happy to clean it up